kubernetes/cache.go

```diff
@@ -0,0 +1,325 @@
+		GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, bool)
+	}
+
+	controllerImpl struct {
+		stopChan     chan struct{}
+		syncCount    int
+		maxSyncCount int
+		controllers  map[string]cache.SharedIndexInformer
+	}
+)
+
+func newCacheController(clientset kube.Interface, refreshDuration time.Duration, stopChan chan struct{}) cacheController {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	controllers := make(map[string]cache.SharedIndexInformer)
+
+	controllers["Pod"] = sharedInformers.Core().V1().Pods().Informer()
+	controllers["ReplicationController"] = sharedInformers.Core().V1().ReplicationControllers().Informer()
+	controllers["Deployment"] = sharedInformers.Apps().V1beta1().Deployments().Informer()
+	controllers["ReplicaSet"] = sharedInformers.Apps().V1beta2().ReplicaSets().Informer()
+	controllers["StatefulSet"] = sharedInformers.Apps().V1beta2().StatefulSets().Informer()
+	controllers["Job"] = sharedInformers.Batch().V1().Jobs().Informer()
+	controllers["CronJob"] = sharedInformers.Batch().V1beta1().CronJobs().Informer()
+	controllers["Service"] = sharedInformers.Core().V1().Services().Informer()
+	controllers["Endpoints"] = sharedInformers.Core().V1().Endpoints().Informer()
+
+	return &controllerImpl{
+		stopChan:     stopChan,
+		controllers:  controllers,
+		syncCount:    0,
+		maxSyncCount: 20, // Move this to config ? or this constant is good enough ?
+	}
+}
+
+func (c *controllerImpl) Run(stop <-chan struct{}) {
+	for _, cn := range c.controllers {
+		go cn.Run(stop)
+	}
+	<-stop
+}
+
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Is there connection with the K8S backend ?")
+		c.StopControlChannel()
+		return false
+	}
+	hasSynced := true
+	for _, cn := range c.controllers {
+		hasSynced = hasSynced && cn.HasSynced()
+	}
+	if hasSynced {
+		c.syncCount = 0
+	} else {
+		c.syncCount = c.syncCount + 1
+	}
+	return hasSynced
+}
+
+func (c *controllerImpl) StopControlChannel() {
+	if c.stopChan != nil {
+		close(c.stopChan)
+		c.stopChan = nil
+	}
+}
+
+func (c *controllerImpl) GetCronJobs(namespace string) ([]batch_v1beta1.CronJob, bool) {
+	indexer := c.controllers["CronJob"].GetIndexer()
+	cronjobs, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(cronjobs) > 0 {
+		_, ok := cronjobs[0].(*batch_v1beta1.CronJob)
+		if !ok {
+			return nil, false
+		}
+		nsCronjobs := make([]batch_v1beta1.CronJob, len(cronjobs))
+		for i, cronjob := range cronjobs {
+			nsCronjobs[i] = *(cronjob.(*batch_v1beta1.CronJob))
+		}
+		return nsCronjobs, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetDeployment(namespace, name string) (*v1beta1.Deployment, bool) {
+	indexer := c.controllers["Deployment"].GetIndexer()
+	deps, exist, err := indexer.GetByKey(namespace + "/" + name) // indexer.ByIndex("nsname", namespace + "#" + name)
```

Comment:
What's the commented out code here? Another way to get same result, or just a relic of your investigations? :)

Reply:
Ah, left code there ! Forgot to clean, perhaps there are a couple of unnecessary logging out there.
Thanks